cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -227,3 +229,52 @@ def flip_inversion(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
     return transformer_primitives.map_operations(
         circuit, flip_inversion, deep=context.deep if context else True, tags_to_ignore=ignored
     ).unfreeze()
+
+
+class _CX(ops.Gate):
+    """A CX gate generalized for qudits.
+
+    This represents a Controlled-NOT gate for qudits, in the following sense. If the target is
+    zero, then it becomes the value of the control. If the target is equal to the control, then
+    it becomes zero. All other cases do not affect the target. The first rule is the only one
+    required for measurement-deferral purposes: the ancilla qudit representing the creg is always
+    zero at the time of measurement. The remaining rules are for symmetry.
+
+    |k0> -> |kk>
+    |kk> -> |k0>
+    |kj> -> |kj> otherwise
+
+    The unitary is formed directly from these rules, thus fully defining behavior in the presence
+    of superposition and entanglement. This definition preserves 2D CX behavior, such as
+    CX∘CX == I, and CX∘XC∘CX == SWAP.
+
+    Note that this is explicitly different from a controlled multidimensional X gate. This is easy
+    to see, in that the latter is a STEP gate, allowing the qudit value to increase by at most one.
```

Comment:
FYI: This appears to describe a gate acting on a qubit (d=2) and a qudit (arbitrary d). This is different from the modular addition above.